src/Microsoft.DotNet.Build.Tasks/AddDependenciesToProjectJson.cs

```diff
@@ -171,34 +238,33 @@ private JObject GenerateDependencies(JObject projectJsonRoot, ITaskItem[] extern
+                    ITaskItem package = GetPackageInformation(property.Name, packageInformation);
 
-                    // if the package is in the external packages list then we'll
-                    // need to replace the version with the version from the list
-                    string externalVersion = null;
-                    if (externalPackageVersions != null)
-                        externalVersion = externalPackageVersions.FirstOrDefault(epv => epv.ItemSpec.Equals(property.Name, StringComparison.OrdinalIgnoreCase))?.GetMetadata("Version");
-
-                    if (updateDependency.Success || externalVersion != null)
+                    if (package != null)
                     {
                         NuGetVersion nuGetVersion;
-                        if (externalVersion == null)
+                        if (NuGetVersion.TryParse(property.Value.ToString(), out nuGetVersion))
                         {
-                            if (NuGetVersion.TryParse(property.Value.ToString(), out nuGetVersion))
+                            NuGetVersion dependencyVersion = nuGetVersion;
+                            string ver = null;
+
+                            // a package version was provided, use its version information.
+                            if (package != null)
+                            {
+                                ver = package.GetMetadata("Version");
+                            }
+                            else
                             {
-                                Match m = _versionStructureRegex.Match(nuGetVersion.ToString());
+                                ver = string.Join(".", dependencyVersion.Major, dependencyVersion.Minor, dependencyVersion.Patch);
+                            }
 
-                                if (m.Success)
-                                {
-                                    NuGetVersion dependencyVersion = nuGetVersion;
-                                    nuGetVersion = NuGetVersion.Parse(string.Join(".", dependencyVersion.Major, dependencyVersion.Minor, dependencyVersion.Patch) + "-" + PackageBuildNumberOverride);
-                                }
+                            string releaseVersion = package.GetMetadata("ReleaseVersion");
+                            // we have package information, so use that.
+                            if (!string.IsNullOrWhiteSpace(releaseVersion))
+                            {
+                                ver = ver + "-" + releaseVersion;
                             }
-                        }
-                        else
-                        {
-                            nuGetVersion = NuGetVersion.Parse(externalVersion);
+                            nuGetVersion = NuGetVersion.Parse(ver);
                         }
 
                         // Only add the original dependency if it wasn't passed as an AdditionalDependency, ie. AdditionalDependencies may override dependencies in project.json
```

Comment:
orthogonal but it might be worth creating a dictionary from the AdditionalDependencies list at some point for efficiency. If you stick with Linq there is also an Any() extension method which does exactly what you are trying to do here without needing to check for null.